File: backend/app/schemas/manual_trading.py
"""手动模拟交易 Pydantic 模型。"""

from datetime import date, datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


class CreateSessionRequest(BaseModel):
    asset_name: str = Field(..., min_length=1, max_length=100)
    start_date: date
    name: Optional[str] = Field(default=None, max_length=100)


class BuyRequest(BaseModel):
    amount: float = Field(..., gt=0)
    price: float = Field(..., gt=0)


class AdvanceRequest(BaseModel):
    step: Literal["day", "week", "month", "year"]


class RevalRequest(BaseModel):
    close_price: float = Field(..., gt=0)


class SessionListItem(BaseModel):
    session_id: str
    name: Optional[str]
    asset_name: str
    start_date: date
    current_date: date
    position_value: float
    total_invested: float
    total_pnl: float
    status: str
    awaiting_reval: bool
    created_at: datetime


class SessionListResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: list[SessionListItem]


class SessionDetailResponse(BaseModel):
    session_id: str
    name: Optional[str]
    asset_name: str
    start_date: date
    current_date: date
    reference_price: Optional[float]
    position_value: float
    total_invested: float
    total_pnl: float
    total_pnl_pct: Optional[float]
    total_trading_days: Optional[int]
    status: str
    awaiting_reval: bool
    end_date: Optional[date]
    first_operation_date: Optional[date]
    created_at: datetime
    updated_at: datetime


class AdvanceResponse(BaseModel):
    session_id: str
    current_date: date
    awaiting_reval: bool
    step: str


class OperationItem(BaseModel):
    id: int
    op_type: str
    op_type_label: str
    op_date: date
    price: Optional[float]
    buy_amount: Optional[float]
    advance_step: Optional[str]
    position_before: float
    position_after: float
    segment_pnl: Optional[float]
    days_since_prev: Optional[int]


class OperationsResponse(BaseModel):
    session_id: str
    total_pnl: float
    total_trading_days: Optional[int]
    items: list[OperationItem]

File: frontend/src/api/manualTrading.ts
import http from './http'

export interface CreateSessionRequest {
  asset_name: string
  start_date: string
  name?: string
}

export interface SessionListItem {
  session_id: string
  name: string | null
  asset_name: string
  start_date: string
  current_date: string
  position_value: number
  total_invested: number
  total_pnl: number
  status: string
  awaiting_reval: boolean
  created_at: string
}

export interface SessionListResponse {
  total: number
  page: number
  page_size: number
  items: SessionListItem[]
}

export interface SessionDetail {
  session_id: string
  name: string | null
  asset_name: string
  start_date: string
  current_date: string
  reference_price: number | null
  position_value: number
  total_invested: number
  total_pnl: number
  total_pnl_pct: number | null
  total_trading_days: number | null
  status: string
  awaiting_reval: boolean
  end_date: string | null
  first_operation_date: string | null
  created_at: string
  updated_at: string
}

export interface AdvanceResponse {
  session_id: string
  current_date: string
  awaiting_reval: boolean
  step: string
}

export interface OperationItem {
  id: number
  op_type: string
  op_type_label: string
  op_date: string
  price: number | null
  buy_amount: number | null
  advance_step: string | null
  position_before: number
  position_after: number
  segment_pnl: number | null
  days_since_prev: number | null
}

export interface OperationsResponse {
  session_id: string
  total_pnl: number
  total_trading_days: number | null
  items: OperationItem[]
}

export type AdvanceStep = 'day' | 'week' | 'month' | 'year'

export async function createSession(body: CreateSessionRequest): Promise<SessionDetail> {
  const { data } = await http.post<SessionDetail>('/manual-trading/sessions', body)
  return data
}

export async function listSessions(params?: {
  status?: string
  page?: number
  page_size?: number
}): Promise<SessionListResponse> {
  const { data } = await http.get<SessionListResponse>('/manual-trading/sessions', { params })
  return data
}

export async function getSession(sessionId: string): Promise<SessionDetail> {
  const { data } = await http.get<SessionDetail>(`/manual-trading/sessions/${sessionId}`)
  return data
}

export async function deleteSession(sessionId: string): Promise<void> {
  await http.delete(`/manual-trading/sessions/${sessionId}`)
}

export async function buySession(
  sessionId: string,
  amount: number,
  price: number,
): Promise<SessionDetail> {
  const { data } = await http.post<SessionDetail>(`/manual-trading/sessions/${sessionId}/buy`, {
    amount,
    price,
  })
  return data
}

export async function advanceSession(
  sessionId: string,
  step: AdvanceStep,
): Promise<AdvanceResponse> {
  const { data } = await http.post<AdvanceResponse>(
    `/manual-trading/sessions/${sessionId}/advance`,
    { step },
  )
  return data
}

export async function revalSession(sessionId: string, closePrice: number): Promise<SessionDetail> {
  const { data } = await http.post<SessionDetail>(`/manual-trading/sessions/${sessionId}/reval`, {
    close_price: closePrice,
  })
  return data
}

export async function endSession(sessionId: string): Promise<SessionDetail> {
  const { data } = await http.post<SessionDetail>(`/manual-trading/sessions/${sessionId}/end`)
  return data
}

export async function listOperations(sessionId: string): Promise<OperationsResponse> {
  const { data } = await http.get<OperationsResponse>(
    `/manual-trading/sessions/${sessionId}/operations`,
  )
  return data
}
